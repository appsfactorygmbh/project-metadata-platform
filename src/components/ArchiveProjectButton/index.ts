import ArchiveProjectButton from './ArchiveProjectButton.vue';

export { ArchiveProjectButton };
