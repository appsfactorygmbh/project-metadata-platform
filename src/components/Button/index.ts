import FloatingButton from './FloatingButton.vue';
import FloatingButtonGroup from './FloatingButtonGroup.vue';

export { FloatingButton, FloatingButtonGroup };
