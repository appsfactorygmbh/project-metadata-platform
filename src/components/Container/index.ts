import BackgroundContainer from './BackgroundContainer.vue';

export { BackgroundContainer };
