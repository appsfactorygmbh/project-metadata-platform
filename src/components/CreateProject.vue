<script lang="ts" setup>
import { ref } from 'vue';
import { PlusOutlined } from '@ant-design/icons-vue';
import { FontColorsOutlined, ShoppingOutlined, TeamOutlined, BankOutlined, UserOutlined } from '@ant-design/icons-vue';
const open = ref<boolean>(false);

const projectName = ref<string>('');
const businessUnit = ref<string>('');
const teamNumber = ref<string>('');
const department = ref<string>('');
const clientName = ref<string>('');

const projectNameStatus = ref<string>('');
const businessUnitStatus = ref<string>('');
const teamNumberStatus = ref<string>('');
const departmentStatus = ref<string>('');
const clientNameStatus = ref<string>('');

const showModal = () => {
  open.value = true;
};

const validateField = (fieldValue: string, fieldStatus: { value: string }) => {
  if (!fieldValue) {
    fieldStatus.value = 'error';
  } else {
    fieldStatus.value = '';
  }
}

const handleOk = () => {
  validateField(projectName.value, projectNameStatus);
  validateField(businessUnit.value, businessUnitStatus);
  validateField(teamNumber.value, teamNumberStatus);
  validateField(department.value, departmentStatus);
  validateField(clientName.value, clientNameStatus);

  if(projectName.value && businessUnit.value && teamNumber.value && department.value && clientName.value){
    open.value = false;
  }

};
</script>

<template>
  <div>
    <a-float-button @click="showModal">
      <template #icon>
        <PlusOutlined />
      </template>
    </a-float-button>
    <a-modal v-model:open="open" width="400px" title="Create Project" @ok="handleOk">
      <a-space direction="vertical" class="space">
        <a-input id="projectNameField" class="inputField" :status="projectNameStatus" v-model:value="projectName" placeholder="Project Name">
          <template #prefix>
            <FontColorsOutlined />
          </template>
        </a-input>
        <a-input id="businessUnitField" class="inputField" :status="businessUnitStatus" v-model:value="businessUnit" placeholder="Business Unit">
          <template #prefix>
            <ShoppingOutlined />
          </template>
        </a-input>
        <a-input id="teamNumberField" class="inputField" :status="teamNumberStatus" v-model:value="teamNumber" placeholder="Team Number">
          <template #prefix>
            <TeamOutlined />
          </template>
        </a-input>
        <a-input id="departmentField" class="inputField" :status="departmentStatus" v-model:value="department" placeholder="Department">
          <template #prefix>
            <BankOutlined />
          </template>
        </a-input>
        <a-input id="clientNameField" class="inputField" :status="clientNameStatus" v-model:value="clientName" placeholder="Client Name">
          <template #prefix>
            <UserOutlined />
          </template>
        </a-input>
      </a-space>
    </a-modal>
  </div>
</template>

<style scoped lang="scss">
.space {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  & > * {
    width: 100%;
  }
}

.inputField {
  width: 90%;


}
</style>
