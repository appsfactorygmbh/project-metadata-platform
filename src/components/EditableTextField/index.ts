import EditableTextField from './EditableTextField.vue';
import EmailInputField from './InputFields/EmailInputField.vue';
import PasswordInputField from './InputFields/PasswordInputField.vue';
import ProjectInformationInputField from './InputFields/ProjectInformationInputField.vue';
import ProjectInformationSelectField from './InputFields/ProjectInformationSelectField.vue';
import TeamNameInputField from './InputFields/TeamNameInputField.vue';

export {
  EditableTextField,
  EmailInputField,
  PasswordInputField,
  TeamNameInputField,
  ProjectInformationInputField,
  ProjectInformationSelectField,
};
