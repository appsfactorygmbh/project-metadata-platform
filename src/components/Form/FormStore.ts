import { defineStore } from 'pinia';
import { ref, type Ref } from 'vue';
import type { FormType } from './types';
import type { FieldData } from 'ant-design-vue/es/form/interface';
import { Form } from 'ant-design-vue';
import type { Rule } from 'ant-design-vue/es/form';
import type { ArgsType } from '@/models/utils';

const { useForm } = Form;

type FieldValue = string | number | boolean | Date | undefined;

type FieldRecord<T> = Record<string, FieldValue | T>;

type AnyArray = Array<FieldRecord<FieldValue>>;

type FormState = FieldRecord<AnyArray>;

export type RulesObject<T> = Record<keyof T, Rule[]>;

type FormStoreState<T extends FormState> = {
  form: Ref<FormType>;
  modelRef: Ref<T>;
  rulesRef: Ref<RulesObject<T>>;
  onSubmit?: (values: T) => void;
  options?: ArgsType<typeof useForm>[2];
};

export const useFormStore = <T extends FormState>(
  name: string,
  form?: FormType,
  modelRef?: T,
  rulesRef?: RulesObject<T>,
  options?: ArgsType<typeof useForm>[2],
) =>
  defineStore(`${name}_form`, {
    state: (): FormStoreState<T> => ({
      modelRef: modelRef
        ? (ref<T>(modelRef) as Ref<T>)
        : (ref<T>({} as T) as Ref<T>),
      rulesRef: rulesRef
        ? (ref<RulesObject<T>>(rulesRef) as Ref<RulesObject<T>>)
        : (ref<RulesObject<T>>({} as RulesObject<T>) as Ref<RulesObject<T>>),
      form: form
        ? (ref<FormType>(form) as Ref<FormType>)
        : (ref<FormType>({} as FormType) as Ref<FormType>),
      options: options,
    }),
    actions: {
      setOnSubmit(onSubmit: (values: T) => void) {
        this.onSubmit = onSubmit;
      },
      setForm(form: FormType) {
        this.form = form;
      },
      updateForm() {
        this.form = useForm(this.modelRef, this.rulesRef, this.options);
      },
      setModel(modelRef: T) {
        this.modelRef = modelRef;
        this.updateForm();
      },
      setRules(rulesRef: RulesObject<T>) {
        this.rulesRef = rulesRef;
        this.updateForm();
      },
      setOptions(options: ArgsType<typeof useForm>[2]) {
        this.options = options;
        this.updateForm();
      },
      updateField(key: keyof T, value: T[keyof T]) {
        this.modelRef[key] = value;
        this.updateForm();
      },
      updateFields(fields: FieldData[]) {
        console.log('fields', fields);
        fields.forEach(({ name, value }) => {
          this.modelRef[name as keyof T] = value as T[keyof T];
        });
        this.updateForm();
      },
      resetFields() {
        this.modelRef = {} as T;
        this.updateForm();
      },
      getFieldValue(key: keyof T): T[keyof T] {
        return this.modelRef[key];
      },
      validate(): Promise<void> {
        this.updateForm();
        return this.form.validate();
      },
      clearValidate() {
        this.form.clearValidate();
      },
      async submit() {
        if (!this.rulesRef) {
          console.warn('Form rules are not defined before submitting.');
        }
        return this.form.validate().then(() => {
          if (this.onSubmit) {
            return this.onSubmit(this.modelRef);
          }
        });
      },
    },
    getters: {
      getFieldsValue(): T {
        return this.modelRef;
      },
      formIsDefined(): boolean {
        return !!this.form.validate;
      },
      getForm(): FormType {
        return this.form;
      },
      getModel(): T {
        return this.modelRef;
      },
      validateInfos(): FormType['validateInfos'] {
        return this.form.validateInfos;
      },
    },
  })();
export type FormStore<T extends FormState = FormState> = ReturnType<
  typeof useFormStore<T>
>;
