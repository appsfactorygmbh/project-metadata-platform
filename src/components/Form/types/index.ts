export type { FormType, FormSubmitType } from './FormTypes';
