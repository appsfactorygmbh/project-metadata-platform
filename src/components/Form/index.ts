import type { FormType, FormSubmitType } from './types';
import { useFormStore, type FormStore } from './FormStore';

export type { FormType, FormSubmitType, FormStore };
export { useFormStore };
