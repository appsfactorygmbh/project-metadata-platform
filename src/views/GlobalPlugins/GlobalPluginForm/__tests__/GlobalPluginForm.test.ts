import { describe, it, expect, expectTypeOf } from 'vitest';
import { mount, VueWrapper } from '@vue/test-utils';
import { GlobalPluginForm, type GlobalPluginFormData } from '..';
import { useFormStore } from '@/components/Form';
import { createPinia, setActivePinia } from 'pinia';

const testForm: GlobalPluginFormData = {
  pluginName: 'testPlugin',
  keys: [
    {
      key: 1231231203,
      value: 'testValue',
    },
    {
      key: 1231231203,
      value: 'testValue2',
    },
    {
      key: 1231231203,
      value: 'testValue3',
    },
  ],
};

describe('GlobalPluginForm.vue', () => {
  let wrapper: VueWrapper;
  let formStore: ReturnType<typeof useFormStore>;

  beforeEach(() => {
    setActivePinia(createPinia());
    formStore = useFormStore('testForm');
    formStore.resetFields();

    wrapper = mount(GlobalPluginForm, {
      props: {
        formStore,
        initialValues: testForm,
      },
    });
  });

  it('should export a valid component', () => {
    expect(GlobalPluginForm).not.toBeUndefined();
    expectTypeOf(GlobalPluginForm).toMatchTypeOf<import('vue').Component>();
  });

  it("should render an empty form if there's no initialValues", () => {
    setActivePinia(createPinia());
    const formStore = useFormStore('testForm');

    wrapper = mount(GlobalPluginForm, {
      props: {
        formStore,
        initialValues: { pluginName: '', keys: [] },
      },
    });
  });

  it('should render a form', () => {
    expect(wrapper.find('form').exists()).toBe(true);
  });

  it('should render the correct number of input fields', () => {
    expect(wrapper.findAll('input').length).toBe(4);
  });

  it("should effect the form store's values", async () => {
    const input = wrapper.find('input');
    await input.setValue('test');
    expect(formStore.getFieldValue('pluginName')).toBe('test');
  });

  it('should reset the form', async () => {
    const input = wrapper.find('input');
    await input.setValue('test');
    expect(formStore.getFieldValue('pluginName')).toBe('test');

    await formStore.resetFields();
    expect(formStore.getFieldValue('pluginName')).toBe(undefined);
  });

  it("should add key fields when 'Add Key' is clicked", async () => {
    const button = wrapper.find('button');
    const count = wrapper.findAll('input').length;
    await button.trigger('click');
    expect(wrapper.findAll('input').length).toBe(count + 1);
  });

  it.todo("should remove key fields when 'Remove Key' is clicked", async () => {
    const button = wrapper.find('button');
    await button.trigger('click');
    const count = wrapper.findAll('input').length;
    await button.trigger('click');
    expect(wrapper.findAll('input').length).toBe(count - 1);
  });

  it('should validate the form if pluginName is not set', async () => {
    const input = wrapper.find('input');
    await input.setValue('');
    expect(formStore.validate()).rejects.toMatchObject({
      errorFields: [
        {
          errors: ['Please insert the plugin name.'],
          name: 'pluginName',
          warnings: [],
        },
      ],
    });

    await input.setValue('testPlugin');
    expect(formStore.validate()).resolves.toEqual(testForm);
  });

  it.todo('should validate the form if keys are not set', async () => {
    const button = wrapper.find('button');
    await button.trigger('click');
    await button.trigger('click');
    await button.trigger('click');

    expect(formStore.validate()).rejects.toMatchObject({
      errorFields: [
        {
          errors: ['Please insert the key.'],
          name: 'keys[0].key',
          warnings: [],
        },
        {
          errors: ['Please insert the value.'],
          name: 'keys[0].value',
          warnings: [],
        },
        {
          errors: ['Please insert the key.'],
          name: 'keys[1].key',
          warnings: [],
        },
        {
          errors: ['Please insert the value.'],
          name: 'keys[1].value',
          warnings: [],
        },
        {
          errors: ['Please insert the key.'],
          name: 'keys[2].key',
          warnings: [],
        },
        {
          errors: ['Please insert the value.'],
          name: 'keys[2].value',
          warnings: [],
        },
      ],
    });

    const inputs = wrapper.findAll('input');
    await inputs[0].setValue('1231231203');
    await inputs[1].setValue('testValue');
    await inputs[2].setValue('1231231203');
    await inputs[3].setValue('testValue2');
    await inputs[4].setValue('1231231203');
    await inputs[5].setValue('testValue3');

    expect(formStore.validate()).resolves.toEqual(formStore.getFieldsValue);
  });
});
