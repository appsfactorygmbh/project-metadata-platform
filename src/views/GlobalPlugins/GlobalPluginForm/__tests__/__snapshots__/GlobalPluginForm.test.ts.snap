// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GlobalPluginForm.vue > should render correctly 1`] = `
"<form class="ant-form ant-form-horizontal css-dev-only-do-not-override-16pw25h"><input type="text" lazy="true" placeholder="Plugin Name" id="form_item_pluginName" class="ant-input css-dev-only-do-not-override-16pw25h inputField" value="testPlugin">
  <div class="ant-form-item css-dev-only-do-not-override-16pw25h">
    <div class="ant-row ant-form-row css-dev-only-do-not-override-16pw25h">
      <div class="ant-col ant-col-xs-24 ant-col-sm-4 ant-form-item-label css-dev-only-do-not-override-16pw25h"><label for="form_item_keys_0_value" class="ant-form-item-required" title="ProjectKeys">ProjectKeys</label></div>
      <div class="ant-col ant-col-xs-24 ant-col-sm-20 ant-form-item-control css-dev-only-do-not-override-16pw25h">
        <div class="ant-form-item-control-input">
          <div class="ant-form-item-control-input-content"><input type="text" lazy="true" placeholder="add the plugin key here" id="form_item_keys_0_value" style="width: 60%; margin-right: 8px;" class="ant-input css-dev-only-do-not-override-16pw25h" value="testValue"><span role="img" aria-label="minus-circle" class="anticon anticon-minus-circle dynamic-delete-button" tabindex="-1"><svg focusable="false" data-icon="minus-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M696 480H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"></path><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"></path></svg><!----></span></div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <!---->
  </div>
  <div class="ant-form-item css-dev-only-do-not-override-16pw25h">
    <div class="ant-row ant-form-row css-dev-only-do-not-override-16pw25h">
      <!---->
      <div class="ant-col ant-col-xs-24 ant-col-xs-offset-0 ant-col-sm-20 ant-col-sm-offset-4 ant-form-item-control css-dev-only-do-not-override-16pw25h">
        <div class="ant-form-item-control-input">
          <div class="ant-form-item-control-input-content"><input type="text" lazy="true" placeholder="add the plugin key here" id="form_item_keys_1_value" style="width: 60%; margin-right: 8px;" class="ant-input css-dev-only-do-not-override-16pw25h" value="testValue2"><span role="img" aria-label="minus-circle" class="anticon anticon-minus-circle dynamic-delete-button" tabindex="-1"><svg focusable="false" data-icon="minus-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M696 480H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"></path><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"></path></svg><!----></span></div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <!---->
  </div>
  <div class="ant-form-item css-dev-only-do-not-override-16pw25h">
    <div class="ant-row ant-form-row css-dev-only-do-not-override-16pw25h">
      <!---->
      <div class="ant-col ant-col-xs-24 ant-col-xs-offset-0 ant-col-sm-20 ant-col-sm-offset-4 ant-form-item-control css-dev-only-do-not-override-16pw25h">
        <div class="ant-form-item-control-input">
          <div class="ant-form-item-control-input-content"><input type="text" lazy="true" placeholder="add the plugin key here" id="form_item_keys_2_value" style="width: 60%; margin-right: 8px;" class="ant-input css-dev-only-do-not-override-16pw25h" value="testValue3"><span role="img" aria-label="minus-circle" class="anticon anticon-minus-circle dynamic-delete-button" tabindex="-1"><svg focusable="false" data-icon="minus-circle" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M696 480H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h368c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"></path><path d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"></path></svg><!----></span></div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <!---->
  </div>
  <div class="ant-form-item css-dev-only-do-not-override-16pw25h">
    <div class="ant-row ant-form-row css-dev-only-do-not-override-16pw25h">
      <!---->
      <div class="ant-col ant-col-xs-24 ant-col-xs-offset-0 ant-col-sm-20 ant-col-sm-offset-4 ant-form-item-control css-dev-only-do-not-override-16pw25h">
        <div class="ant-form-item-control-input">
          <div class="ant-form-item-control-input-content"><button style="width: 60%;" class="css-dev-only-do-not-override-16pw25h ant-btn ant-btn-dashed" type="button">
              <transition-stub name="ant-btn-loading-icon-motion" appear="false" persisted="false" css="true">
                <!---->
              </transition-stub><span role="img" aria-label="plus" class="anticon anticon-plus"><svg focusable="false" data-icon="plus" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"></path><path d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"></path></svg><!----></span><span>Add ProjectKey</span>
            </button></div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <!---->
  </div>
</form>
<contextholder></contextholder>"
`;
